larpix/larpix_hpt.py

```diff
@@ -0,0 +1,133 @@
+from __future__ import absolute_import
+
+from math import fmod
+
+def get_ns(time_in_sec):
+    return int(fmod(time_in_sec * 1e9, 1e9))
+
+def get_s(time_in_sec):
+    return int(time_in_sec)
+
+class larpix_hpt(object):
+    '''Class for precision time'''
+    larpix_clk_freq = 5e6 #Hz
+    larpix_offset_d = 2**24 #clk cycles
+    offset = {}
+    ref_time = None
+    prev_time = {}
+
+    def __init__(self, s=0, ns=0, cpu_time=0, adc_time=0):
+        '''Constructor'''
+        self.s = int(s)
+        self.ns = int(fmod(ns,1e9))
+        self.cpu_time = cpu_time
+        self.adc_time = int(adc_time)
+
+    def __str__(self):
+        string = 's: %d' % self.s
+        string += ' ns: %d' % self.ns
+        string += ' cpu_time: %f' % self.cpu_time
+        string += ' adc_time: %d' % self.adc_time
+        return string
+
+    def __repr__(self):
+        return str(self)
+
+    def __eq__(self, hpt):
+        return self.s == hpt.s and self.ns == hpt.ns
+
+    def __gt__(self, hpt):
+        if self.s > hpt.s:
+            return True
+        elif self.s == hpt.s:
+            return self.ns > hpt.ns
+        else:
+            return False
+
+    def __lt__(self, hpt):
+        return not (self > hpt) and not (self == hpt)
+
+    def __add__(self, hpt):
+        sum_ns = int(fmod(self.ns + hpt.ns, 1e9))
+        sum_s = self.s + hpt.s + int((self.ns + hpt.ns) / 1e9)
+        sum_cpu = self.cpu_time + hpt.cpu_time
+        sum_adc = self.adc_time + hpt.adc_time
+        return larpix_hpt(sum_s, sum_ns, sum_cpu, sum_adc)
+
+    def __sub__(self, hpt):
+        delta_ns = int(fmod(self.ns - hpt.ns, 1e9))
+        delta_s = self.s - hpt.s + int((self.ns - hpt.ns) / 1e9)
+        delta_cpu = self.cpu_time - hpt.cpu_time
+        delta_adc = self.adc_time - hpt.adc_time
+        return larpix_hpt(delta_s, delta_ns, delta_cpu, delta_adc)
+
+    def __float__(self):
+        return float(self.s + self.ns / 1e9)
+
+    @staticmethod
+    def from_packet(cpu_time=0, packet=None):
+        '''Constructor from packet'''
+        if packet is None:
+            adc_time = 0
+            chip_id = 0
+        else:
+            adc_time = packet.timestamp
+            chip_id = packet.chipid
+        return larpix_hpt.serialized_hpt(cpu_time, adc_time, chip_id)
+
+    @staticmethod
+    def serialized_hpt(cpu_time=0, adc_time=0, chip_id=0):
+        '''Generates next serialized packet from cpu time and adc time'''
```

Comment:
And also a description of the algorithm:

- For the first packet received from a chip, assumes (whatever) and does (whatever)
- For the first packet in a serial_read (but with previous packets already received), estimates the number of missed cycles using the cpu timestamps and then (whatever)
- For packets that share a cpu timestamp (same serial_read), try to guess the number of missed cycles (how exactly?) and then (whatever)